src/adapters/UniswapV3Adapter.sol

```diff
@@ -0,0 +1,125 @@
+// SPDX-License-Identifier: AGPL-v3.0
+
+pragma solidity 0.8.17;
+
+import "../Feedbase.sol";
+
+interface IUniswapV3Pool {
+    function observe(uint32[] calldata secondsAgos) external view
+        returns (
+            int56[] memory tickCumulatives,
+            uint160[] memory secondsPerLiquidityCumulativeX128s
+        );
+}
+
+
+contract UniswapV3Adapter {
+    struct Config {
+        address pool;
+        uint    range;
+        uint    ttl;
+        bool    reverse;
+    }
+    Feedbase immutable fb;
+    mapping(address=>bool) public wards;
+    mapping(bytes32=>Config) public configs;
+    uint constant RAY = 10 ** 27;
+    uint constant X96 = 2 ** 96;
+
+    constructor(Feedbase _fb) {
+        wards[msg.sender] = true;
+        fb = _fb;
+    }
+
+    modifier _ward_ {
+        require(wards[msg.sender], "unwarded sender");
+        _;
+    }
+
+    function ward(address _owner, bool ok) public _ward_ {
+        wards[_owner] = ok;
+    }
+
+    function setConfig(bytes32 tag, Config memory config) public _ward_ {
+        configs[tag] = config;
+    }
+
+    function look(bytes32 tag) public {
+        Config storage config = configs[tag];
+        address apool = config.pool;
+        require(address(0) != apool, "no pool for tag");
+
+        uint32[] memory times = new uint32[](2);
+        uint32 range  = uint32(config.range);
+        require(range > 0, "twap range can't be zero");
+        times[0] = 0;
+        times[1] = range;
+        (int56[] memory cumulatives,) = IUniswapV3Pool(apool).observe(times);
+
+        int   delt         = int(cumulatives[0]) - int(cumulatives[1]);
+        int24 meantick     = int24(delt / int(uint(range)));
+        uint  sqrtPriceX96 = getSqrtRatioAtTick(meantick);
+        uint  priceray     = sqrtPriceX96 ** 2 / (X96) * RAY / (X96);
+        if (config.reverse) {
+            priceray = RAY * RAY / priceray;
+        }
+        fb.push(tag, bytes32(priceray), block.timestamp + config.ttl);
+    }
+
+    /////////////////////////////////////////////////////////////////////////////////
+    // this section copied from UniswapV3 and modified to compile:
+    // https://github.com/Uniswap/v3-core/blob/main/contracts/libraries/TickMath.sol
+    // SPDX-License-Identifier: GPL-2.0-or-later
+
+    /// @dev The minimum tick that may be passed to #getSqrtRatioAtTick computed from log base 1.0001 of 2**-128
+    int24 internal constant MIN_TICK = -887272;
+    /// @dev The maximum tick that may be passed to #getSqrtRatioAtTick computed from log base 1.0001 of 2**128
+    int24 internal constant MAX_TICK = -MIN_TICK;
+
+    /// @dev The minimum value that can be returned from #getSqrtRatioAtTick. Equivalent to getSqrtRatioAtTick(MIN_TICK)
+    uint160 internal constant MIN_SQRT_RATIO = 4295128739;
```

Comment:
SQRT_RATIO constants and one of the ticks unused

Reply:
MAX_TICK depends on MIN_TICK